### ISSUE TITLE
Getting instance of sun.misc.Unsafe class does not work on HPUX operation system

### ISSUE THREAD

**Issue by basovnik** [hazelcast/hazelcast#5518]
See [error report](https://gist.github.com/basovnik/a1c0bb364930d9374e28)

**Comment by pveentjer** [comment]
Should this not be classified as a bug?

**Comment by serkan-ozal** [comment]
Hi @basovnik 
As far as I see from the crash logs, you are using 32 bit JVM (addresses are 4 byte length)
On 32 bit JVM, array base offset is `12 bytes=0x0C` (object header [4 bytes] + class definition pointer [4 bytes] + array length [4 bytes]) 

When I see the crash logs, problematic address is `0x6b3c7aa4` and also on the top of stack, I see this:

```
3dfff5a0:   00000000 00000004 00000000 0000000c
3dfff5b0:   6b3c7a98 550806a8 6b3c7a98 00000000
``` 

`6b3c7a98` is the address of temporary byte[] defined here https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/nio/UnsafeHelper.java#L91 and `0000000c` is the byte array base offset passed to Unsafe here https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/nio/UnsafeHelper.java#L96. So the target address is `6b3c7a98` + `0000000c` = `0x6b3c7aa4` which is not multiply of `8`. So I guess, JVM/HPUX expects `8` byte aligned address for putting `long` value since `long` is `8` bytes sized.

In summary, I thinkthis is not the Hazelcast bug but JVM bug based on **HPUX** operating system about invalid address alignment as mentioned here http://h30499.www3.hp.com/t5/Languages-and-Scripting/BUS-ADRALN-Invalid-address-alignment-while-C-program-execution/td-p/5678259#.VYHui_ntmko

Regards.

**Comment by jerrinot** [comment]
I'm assigning the issue to myself. I discussed it with @serkan-ozal and we came to a conclusion we'd provide a workaround for platforms where aligned access is required.

Many thanks to @basovnik for reporting it!

**Comment by serkan-ozal** [comment]
In addition i advice a new system property such as `hazelcast.disableUnsafe`. If this property is set, we can make unsafe unavailable for our system for handling future unsafe issues. So if user really doesn't need to unsafe (like in this issue),  he/she can simply disable unsafe without waiting our fix/patch

**Comment by Donnerbart** [comment]
:+1: for the system property. Who knows on which systems we might experience Unsafe problems in the future. Shouldn't harm to have a switch for this.

**Comment by jerrinot** [comment]
#5540 disables Unsafe on platforms where unaligned platforms is not supported. It introduces a white-list of platforms where unaligned access is known to work.

It also defines a new system property `hazelcast.unsafe.mode` you can set to:
- `disabled` to explicitly disable Unsafe usage 
- `enforced` to enforce it even when running on an unsupported platforms

It's meant to be a workaround for situations when the auto-detection mechanism fails.
